hashicorp/go-tfe
# StateVersionOutputs

## Description

The `Value` field in `StateVersionOutput` can be any number of types, and is currently restricted to being only a string. This issue has been brought up before (#231 #158). 

A recent endpoint was added to [StateVersions to retrieve all the outputs for a given StateVersion](https://www.terraform.io/docs/cloud/api/state-versions.html#list-state-version-outputs). This is different than adding `outputs` to an [include query](https://www.terraform.io/docs/cloud/api/state-versions.html#available-related-resources) due to how we handle permissions. This new endpoint follows _Read state outputs_ in the [Workspace Permissions](https://www.terraform.io/docs/cloud/users-teams-organizations/permissions.html#workspace-permissions).

The value is now an `interface{}` and can be type cast to any other type. See the `TestStateVersionOutputs` for some examples.

## Testing plan

1. Create a workspace and a run that produces outputs
2. Use this new handler and pass in the state version ID.
3. Ensure the outputs are all present and the Value for each one is also present. You can do `fmt.Printf("%v", output.Value)` to see it.

## External links

- [StateVersionOutputs API documentation](https://www.terraform.io/docs/cloud/api/state-versions.html#list-state-version-outputs)
- [jsonapi change](https://github.com/hashicorp/jsonapi/pull/11)
- Issue #231 
- Issue #158 

## Output from tests

```
go-tfe % TFE_TOKEN=<token> TFE_ADDRESS=https://tfe-zone-4c292d5b.ngrok.io/ go test -v -run TestState
=== RUN   TestStateVersionOutputsRead
    --- PASS: TestStateVersionsList/without_list_options (0.27s)                                                                                                                                                                                                                                                                        [168/5345]
    --- PASS: TestStateVersionsList/with_list_options (0.36s)
    --- PASS: TestStateVersionsList/without_an_organization (0.00s)
    --- PASS: TestStateVersionsList/without_a_workspace (0.00s)
=== RUN   TestStateVersionsCreate
=== RUN   TestStateVersionsCreate/with_valid_options
=== RUN   TestStateVersionsCreate/with_the_force_flag_set
=== RUN   TestStateVersionsCreate/with_a_run_to_associate_with
    state_version_test.go:196: This can only be tested with the run specific token
=== RUN   TestStateVersionsCreate/without_md5_hash
=== RUN   TestStateVersionsCreate/withous_serial
=== RUN   TestStateVersionsCreate/without_state
=== RUN   TestStateVersionsCreate/with_invalid_workspace_id
--- PASS: TestStateVersionsCreate (4.19s)
    --- PASS: TestStateVersionsCreate/with_valid_options (1.07s)
    --- PASS: TestStateVersionsCreate/with_the_force_flag_set (1.58s)
    --- SKIP: TestStateVersionsCreate/with_a_run_to_associate_with (0.00s)
    --- PASS: TestStateVersionsCreate/without_md5_hash (0.00s)
    --- PASS: TestStateVersionsCreate/withous_serial (0.00s)
    --- PASS: TestStateVersionsCreate/without_state (0.00s)
    --- PASS: TestStateVersionsCreate/with_invalid_workspace_id (0.00s)
=== RUN   TestStateVersionsRead
=== RUN   TestStateVersionsRead/when_the_state_version_exists
=== RUN   TestStateVersionsRead/when_the_state_version_does_not_exist
=== RUN   TestStateVersionsRead/with_invalid_state_version_id
--- PASS: TestStateVersionsRead (3.43s)
    --- PASS: TestStateVersionsRead/when_the_state_version_exists (0.31s)
    --- PASS: TestStateVersionsRead/when_the_state_version_does_not_exist (0.47s)
    --- PASS: TestStateVersionsRead/with_invalid_state_version_id (0.00s)
=== RUN   TestStateVersionsReadWithOptions
=== RUN   TestStateVersionsReadWithOptions/when_the_state_version_exists
--- PASS: TestStateVersionsReadWithOptions (3.03s)
    --- PASS: TestStateVersionsReadWithOptions/when_the_state_version_exists (0.23s)
=== RUN   TestStateVersionsCurrent
=== RUN   TestStateVersionsCurrent/when_a_state_version_exists
=== RUN   TestStateVersionsCurrent/when_a_state_version_does_not_exist
=== RUN   TestStateVersionsCurrent/with_invalid_workspace_id
--- PASS: TestStateVersionsCurrent (3.99s)
    --- PASS: TestStateVersionsCurrent/when_a_state_version_exists (0.22s)
    --- PASS: TestStateVersionsCurrent/when_a_state_version_does_not_exist (0.30s)
    --- PASS: TestStateVersionsCurrent/with_invalid_workspace_id (0.00s)
=== RUN   TestStateVersionsCurrentWithOptions
=== RUN   TestStateVersionsCurrentWithOptions/when_the_state_version_exists
--- PASS: TestStateVersionsCurrentWithOptions (3.27s)
    --- PASS: TestStateVersionsCurrentWithOptions/when_the_state_version_exists (0.34s)
=== RUN   TestStateVersionsDownload
=== RUN   TestStateVersionsDownload/when_the_state_version_exists
=== RUN   TestStateVersionsDownload/when_the_state_version_does_not_exist
=== RUN   TestStateVersionsDownload/with_an_invalid_url
--- PASS: TestStateVersionsDownload (3.17s)
    --- PASS: TestStateVersionsDownload/when_the_state_version_exists (0.09s)
    --- PASS: TestStateVersionsDownload/when_the_state_version_does_not_exist (0.09s)
    --- PASS: TestStateVersionsDownload/with_an_invalid_url (0.42s)
=== RUN   TestStateVersionOutputsList
--- PASS: TestStateVersionOutputsList (3.62s)
PASS
ok      github.com/hashicorp/go-tfe     33.713
```
